packages/builtin/unsigned.pony

```diff
@@ -385,23 +426,35 @@ primitive ULong is UnsignedInteger[ULong]
 
   fun addc(y: ULong): (ULong, Bool) =>
     ifdef ilp32 or llp64 then
-      @"llvm.uadd.with.overflow.i32"[(ULong, Bool)](this, y)
+      (let r, let o) =
+        @"llvm.uadd.with.overflow.i32"[(U32, Bool)](this.u32(), y.u32())
+      (r.ulong(), o)
     else
-      @"llvm.uadd.with.overflow.i64"[(ULong, Bool)](this, y)
+      (let r, let o) =
+        @"llvm.uadd.with.overflow.i64"[(U64, Bool)](this.u64(), y.u64())
+      (r.ulong(), o)
     end
 
   fun subc(y: ULong): (ULong, Bool) =>
     ifdef ilp32 or llp64 then
-      @"llvm.usub.with.overflow.i32"[(ULong, Bool)](this, y)
+      (let r, let o) =
+        @"llvm.usub.with.overflow.i32"[(U32, Bool)](this.u32(), y.u32())
+      (r.ulong(), o)
```

Comment:
@jemc There are several place like this since LLVM returns a tuple. I'm unsure what the cost of unpacking, casting and packing the tuple again looks like compared to the simple casting that was happening before. Any way I could verify this?

Reply:
I'm quite confident that this will be a zero-cost abstraction.

The tuples are concrete, so there is no cost of boxing and unboxing them from an object descriptor wrapper.

The numeric conversions are of an equivalent LLVM type, so the conversion code is a no-op and the LLVM types of the tuples are equivalent.

We should be good to go, but if you want to verify, you could compile a tiny program with `ponyc --pass=ir` and we could inspect the LLVM IR together over Zulip. Something like this program should work for this purpose:

```mare
use "time"

actor Main
  new create(env: Env) =>
    try
      let long: ULong = env.args(0)?.hash().ulong() % 100
      (let result, let did_underflow) = long.subc(99)
      env.out.print("Result:")
      env.out.print(result.string())
      if did_underflow then env.out.print("(but it underflowed!)") end
    end
```

The reason for using `env.args(0)?.hash()` as the input is to make sure that LLVM cannot optimize away all of the code like it could if we used integer constants.

Reply:
Verified that LLVM optimizes this away, it's identical to what we had before in terms of IR.